extra/modules/greenbids-real-time-data/src/main/java/org/prebid/server/hooks/modules/greenbids/real/time/data/config/DatabaseReaderFactory.java

```diff
@@ -1,52 +1,129 @@
-            try {
-                final URL url = new URL(geoLiteCountryUrl);
-                final Path databasePath = Files.createTempFile("GeoLite2-Country", ".mmdb");
+    private Future<Void> downloadFile(String downloadUrl, String tmpPath) {
+        return vertx.fileSystem().mkdirs(properties.tmpDir)
+                .compose(v -> vertx.fileSystem().open(tmpPath, new OpenOptions()))
+                .compose(tmpFile -> sendHttpRequest(downloadUrl)
+                        .compose(response -> response.pipeTo(tmpFile))
+                        .eventually(v -> tmpFile.close())
+                        .onSuccess(ignored -> logger.info("File downloaded successfully to {}", tmpPath))
+                        .onFailure(error -> logger.error(
+                                "Failed to download file from {} to {}.", downloadUrl, tmpPath, error)));
+    }
+
+    private Future<HttpClientResponse> sendHttpRequest(String url) {
+        final RequestOptions options = new RequestOptions()
+                .setFollowRedirects(true)
+                .setMethod(HttpMethod.GET)
+                .setTimeout(properties.timeoutMs)
+                .setAbsoluteURI(url);
+
+        return vertx.createHttpClient().request(options)
+                .compose(HttpClientRequest::send)
+                .map(this::validateResponse);
+    }
 
-                try (InputStream inputStream = url.openStream();
-                     FileOutputStream outputStream = new FileOutputStream(databasePath.toFile())) {
-                    inputStream.transferTo(outputStream);
+    private HttpClientResponse validateResponse(HttpClientResponse response) {
+        final int statusCode = response.statusCode();
+        if (statusCode != HttpResponseStatus.OK.code()) {
+            throw new PreBidException("Got unexpected response from server with status code %s and message %s"
+                    .formatted(statusCode, response.statusMessage()));
+        }
+        return response;
+    }
+
+    private DatabaseReader extractMMDB(String tarGzPath) {
+        try (GZIPInputStream gis = new GZIPInputStream(Files.newInputStream(Path.of(tarGzPath)));
+                TarArchiveInputStream tarInput = new TarArchiveInputStream(gis)) {
+
+            TarArchiveEntry currentEntry;
+            boolean hasDatabaseFile = false;
+            while ((currentEntry = tarInput.getNextTarEntry()) != null) {
+                if (currentEntry.getName().contains("GeoLite2-Country.mmdb")) {
+                    hasDatabaseFile = true;
+                    break;
                 }
+            }
 
-                databaseReaderRef.set(new DatabaseReader.Builder(databasePath.toFile()).build());
-            } catch (IOException e) {
-                throw new PreBidException("Failed to initialize DatabaseReader from URL", e);
+            if (!hasDatabaseFile) {
+                throw new RuntimeException("GeoLite2-Country.mmdb not found in the archive");
```

Comment:
it's better to use PrebidExceptions and then it's not necessary to catch them above

Reply:
replaced with PreBidException and removed catching in `downloadAndExtract` method